File: polyglot/Decorator/Decorator.scala
// Decorator Design Pattern

// The implementation of the decorator pattern in scala does not follow
// the pattern encountered in other languages, due to scala built in 'traits'

trait Pizza {
  def description : String;
  def cost : Double;
}

class PlainPizza extends Pizza {
  override def description = "Plain pizza"
  override def cost        = 3.0
}

// 'abstract override' is very surprising. The keyword abstract is in fact
// required as the call to 'super' formally looks like a call to the trait Pizza
// However, we shall use these traits with concrete implementation of Pizza.

trait ExtraCheese extends Pizza {
  abstract override def description = super.description + " + extra cheese"
  abstract override def cost        = super.cost + 0.5
}

trait ExtraOlives extends Pizza {
  abstract override def description = super.description + " + extra olives"
  abstract override def cost        = super.cost + 0.7
}

trait ExtraAnchovies extends Pizza {
  abstract override def description = super.description + " + extra anchovies"
  abstract override def cost        = super.cost + 0.8
}

object Decorator {

  def main(args: Array[String]) = {

    val plainPizza  
      = new PlainPizza

    val nicePizza   
      = new PlainPizza with ExtraCheese

    val superPizza  
      = new PlainPizza with ExtraCheese with ExtraOlives

    val ultraPizza  
      = new PlainPizza with ExtraCheese with ExtraOlives with ExtraAnchovies

    println(
      "Plain Pizza: "   +
      "\tCost: "        + plainPizza.cost
    + "\tDescription: " + plainPizza.description)

    println(
      "Nice Pizza: "   +
      "\tCost: "        + nicePizza.cost
    + "\tDescription: " + nicePizza.description)

    println(
      "Super Pizza: "   +
      "\tCost: "        + superPizza.cost
    + "\tDescription: " + superPizza.description)

    println(
      "Ultra Pizza: "   +
      "\tCost: "        + ultraPizza.cost
    + "\tDescription: " + ultraPizza.description)
  }
}

File: polyglot/Decorator/decorator.php
<?php
// Decorator Design Pattern

// The implementation of the decorator pattern in php does not follow
// the pattern encountered in other languages, due to php built in 'traits'

interface Pizza {
  public function description();
  public function cost();
}

class PlainPizza implements Pizza {
  public function description(){ return "Plain Pizza"; }
  public function cost(){ return 3.0; }
}


trait WithExtraCheese {
  public function description(){ 
    return parent::description()." + extra cheese";
  }
  public function cost(){ 
    return parent::cost() + 0.5; 
  }
}

trait WithExtraOlives {
  public function description(){ 
    return parent::description()." + extra olives";
  }
  public function cost(){ 
    return parent::cost() + 0.7; 
  }
}

trait WithExtraAnchovies {
  public function description(){ 
    return parent::description()." + extra anchovies";
  }
  public function cost(){ 
    return parent::cost() + 0.8; 
  }
}

class NicePizza extends PlainPizza {
  use WithExtraCheese;
}

class SuperPizza extends NicePizza {
  use WithExtraOlives;
}

class UltraPizza extends SuperPizza {
  use WithExtraAnchovies;
}

$plainPizza = new PlainPizza();
$nicePizza  = new NicePizza();
$superPizza = new SuperPizza();
$ultraPizza = new UltraPizza();

echo  "Plain Pizza: "   .
      "\tCost: "        . number_format($plainPizza->cost(), 1).
      "\tDescription: " . $plainPizza->description()."\n"; 

echo  "Nice Pizza: "    .
      "\tCost: "        . number_format($nicePizza->cost(), 1).
      "\tDescription: " . $nicePizza->description()."\n"; 

echo  "Super Pizza: "   .
      "\tCost: "        . number_format($superPizza->cost(), 1).
      "\tDescription: " . $superPizza->description()."\n"; 

echo  "Ultra Pizza: "   .
      "\tCost: "        . number_format($ultraPizza->cost(), 1).
      "\tDescription: " . $ultraPizza->description()."\n"; 


?>
